client/client.go

```diff
@@ -0,0 +1,604 @@
+
+[Install]
+WantedBy=sockets.target
+`
+
+func getGoPath() (string, error) {
+	goPathBytes, err := exec.Command("go", "env", "GOPATH").Output()
+	if err != nil {
+		return "", errors.Wrap(err, "Unable to determine GOPATH")
+	}
+	return strings.TrimSpace(string(goPathBytes)), nil
+}
+
+// Setup Installs dependencies
+func Setup(ctx context.Context) error {
+	home := os.Getenv("HOME")
+	if home == "" {
+		return errors.New("HOME is not defined")
+	}
+	home = strings.TrimSpace(home)
+
+	goPath, err := getGoPath()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Installing host dependencies")
+	osprepare.InstallDeps(ctx, ccloudvmDeps, logger{})
+
+	systemdRootPath := filepath.Join(home, ".local/share/systemd/user")
+	err = os.MkdirAll(systemdRootPath, 0700)
+	if err != nil {
+		return errors.Wrap(err, "Unable to create systemd directory")
+	}
+
+	servicePath := filepath.Join(systemdRootPath, "ccloudvm.service")
+	serviceData := fmt.Sprintf(systemdService, goPath)
+	err = ioutil.WriteFile(servicePath, []byte(serviceData), 0600)
+	if err != nil {
+		return errors.Wrap(err, "Unable to write service file")
+	}
+
+	socketPath := filepath.Join(systemdRootPath, "ccloudvm.socket")
+	socketData := fmt.Sprintf(systemdSocket, home)
+	err = ioutil.WriteFile(socketPath, []byte(socketData), 0600)
+	if err != nil {
+		return errors.Wrap(err, "Unable to write service file")
+	}
+
+	err = exec.Command("systemctl", "--user", "daemon-reload").Run()
+	if err != nil {
+		return errors.Wrap(err, "Unable to reload service files")
+	}
+
+	err = exec.Command("systemctl", "--user", "enable", "ccloudvm.socket").Run()
+	if err != nil {
+		return errors.Wrap(err, "Unable to enable ccloudvm.socket")
+	}
+	err = exec.Command("systemctl", "--user", "start", "ccloudvm.socket").Run()
+	if err != nil {
+		return errors.Wrap(err, "Unable to start ccloudvm.socket")
+	}
+
+	return nil
+}
+
+// Teardown disables the ccloudvm service and deletes all existing instances
+func Teardown(ctx context.Context) error {
+	home := os.Getenv("HOME")
+	if home == "" {
+		return errors.New("HOME is not defined")
+	}
+
+	var instances []string
+	err := issueCommand(ctx,
+		func(client *rpc.Client) (int, error) {
+			var id int
+			err := client.Call("ServerAPI.GetInstances", struct{}{}, &id)
+			return id, err
+		},
+		func(client *rpc.Client, id int) error {
+			return client.Call("ServerAPI.GetInstancesResult", id, &instances)
+		})
+	if err == nil {
+		for _, i := range instances {
+			fmt.Printf("Deleting %s\n", i)
+			_ = Delete(ctx, i)
+		}
+	}
+
+	/* There's a tiny chance of a race here.  If someone started creating
+	   a VM in one window and called teardown in another and the instance came into
+	   existence after we called GetInstances but before we shut down the service,
+	   there's a small chance the instance could be leaked. */
+
+	fmt.Println("Removing ccloudvm service")
+
+	_ = exec.Command("systemctl", "--user", "disable", "ccloudvm.service").Run()
```

Comment:
would be nice to have a systemctl struct to wrapper all systemctl commands

```
systemctl.disable("cloudvm.service")
....
```

with this you can also avoid misspell error

Reply:
I think this is a bit overkill.  ccloudvm executes lots of different commands and I don't really want to provide an abstraction for all of these.  Note that if there's a spelling error, things simply won't work so the issue will be detected early.